File: src_erl/main_3.erl
-module(main_3).

-export([main/0]).


-spec main() -> {ok, books_shop:order()} | {error, term()}.
main() ->
    handle_create_order(books_shop:test_data()).


-spec handle_create_order(map()) ->
    {ok, books_shop:order()} | {error, term()}.

handle_create_order(Data0) ->
    try
        Data = books_shop:validate_incoming_data_ex(Data0),
        #{
            <<"cat">> := Cat0,
            <<"address">> := Address0,
            <<"books">> := Books0
        } = Data,
        Cat = books_shop:validate_cat_ex(Cat0),
        Address = books_shop:validate_address_ex(Address0),
        Books = lists:map(
            fun(#{<<"title">> := Title, <<"author">> := Author}) ->
                books_shop:get_book_ex(Title, Author)
            end,
            Books0
        ),
        Order = books_shop:create_order(Cat, Address, Books),
        {ok, Order}
    catch
        throw:Error -> Error
    end.

File: src_ex/Main3.ex
defmodule Main3 do

    @spec main :: {:ok, BookShop.Order.t} | {:error, term}
    def main do
        BookShop.test_data |> handle_create_order
    end


    @spec handle_create_order(map) :: {:ok, BookShop.Order.t} | {:error, term}
    def handle_create_order data0 do
        try do
            data = BookShop.validate_incoming_data_ex data0
            %{
                "cat" => cat0,
                "address" => address0,
                "books" => books0
            } = data
            cat = BookShop.validate_cat_ex cat0
            address = BookShop.validate_address_ex address0
            books = Enum.map books0, fn %{"title" => title, "author" => author } ->
                BookShop.get_book_ex title, author
            end
            order = BookShop.create_order cat, address, books
            {:ok, order}
        catch
            {:error, reason} -> {:error, reason}
        end
    end

end